src/iccpd/src/iccp_csm.c

```diff
@@ -0,0 +1,802 @@
+        && diconn_code->icc_parameter.f_bit == 0x0
+        && diconn_code->icc_parameter.type == TLV_T_DISCONNECT_CODE
+        && ntohs(diconn_code->icc_parameter.len) == (TLV_L_DISCONNECT_CODE)
+        && ntohl(diconn_code->iccp_status_code) == (STATUS_CODE_ICCP_RG_REMOVED))
+    {
+        csm->iccp_info.sender_rg_connect_flag = 0x0;
+        csm->iccp_info.peer_rg_connect_flag = 0x0;
+    }
+}
+
+/* Add received message into message list */
+void iccp_csm_enqueue_msg(struct CSM* csm, struct Msg* msg)
+{
+    ICCHdr* icc_hdr = NULL;
+    NAKTLV* naktlv = NULL;
+    int type = -1;
+    int i = 0;
+
+    if (csm == NULL)
+    {
+        if (msg != NULL)
+            free(msg);
+        return;
+    }
+
+    if (msg == NULL)
+        return;
+
+    icc_hdr = (ICCHdr*)msg->buf;
+
+    *(uint16_t *)icc_hdr = ntohs(*(uint16_t *)icc_hdr);
+
+    if (icc_hdr->ldp_hdr.msg_type == MSG_T_RG_APP_DATA)
+    {
+        app_csm_enqueue_msg(csm, msg);
+    }
+    else if (icc_hdr->ldp_hdr.msg_type == MSG_T_NOTIFICATION)
+    {
+        naktlv = (NAKTLV*)&msg->buf[sizeof(ICCHdr)];
+
+        for (i = 0; i < MAX_MSG_LOG_SIZE; ++i)
+        {
+            if (ntohl(naktlv->rejected_msg_id) == csm->msg_log.msg[i].msg_id)
+            {
+                type = csm->msg_log.msg[i].type;
+                break;
+            }
+        }
+
+        if (type == MSG_T_RG_APP_DATA)
+            app_csm_enqueue_msg(csm, msg);
+        else
+            TAILQ_INSERT_TAIL(&(csm->msg_list), msg, tail);
+    }
+    else
+    {
+        TAILQ_INSERT_TAIL(&(csm->msg_list), msg, tail);
+    }
+}
+
+/* Get received message from message list */
+struct Msg* iccp_csm_dequeue_msg(struct CSM* csm)
+{
+    struct Msg* msg = NULL;
+
+    if (!TAILQ_EMPTY(&(csm->msg_list)))
+    {
+        msg = TAILQ_FIRST(&(csm->msg_list));
+        TAILQ_REMOVE(&(csm->msg_list), msg, tail);
+
+    }
+
+    return msg;
+}
+
+/* Message initialization */
+int iccp_csm_init_msg(struct Msg** msg, char* data, int len)
+{
+    struct Msg* iccp_msg = NULL;
+
+    if (msg == NULL)
+        return -2;
+
+    if (data == NULL || len <= 0)
+        return -1;
+
+    iccp_msg = (struct Msg*)malloc(sizeof(struct Msg));
+    if (iccp_msg == NULL)
+        goto err_ret;
+
+    iccp_msg->buf = (char*)malloc(len);
+    if (iccp_msg->buf == NULL)
+        goto err_ret;
+
+    memcpy(iccp_msg->buf, data, len);
+    iccp_msg->len = len;
+    *msg = iccp_msg;
+
+    return 0;
+
+ err_ret:
+    if (iccp_msg)
+    {
+        if (iccp_msg->buf)
+            free(iccp_msg->buf);
+        free(iccp_msg);
+    }
+
+    return -1;
+}
+
+void iccp_csm_stp_role_count(struct CSM *csm)
+{
+    /* decide the role, lower ip to be active & socket client*/
+    if (csm->role_type == STP_ROLE_NONE)
+    {
+        if (inet_addr(csm->sender_ip) < inet_addr(csm->peer_ip))
+        {
+            /* Active*/
+            ICCPD_LOG_INFO(__FUNCTION__, "Role: [Active]");
+            csm->role_type = STP_ROLE_ACTIVE;
```

Comment:
here iccp role and stp role seems to be same. mlacp_stage_handler() uses this check to iccp role as active/standby. 
does stp and iccp roles are both decided using this variable

Reply:
Yes, iccp role and stp role is same. STP is not supported in SONIC currently.